File: by-python/py-boj/1890-점프.py
def dp(y, x):
    if y == n - 1 and x == n - 1: return 1
    if y >= n or x >= n: return 0

    if memo[y][x] != -1: return memo[y][x]  # arr과 동일한 좌표의 memo 값이 -1이 아니면 이미 지나쳐감
    memo[y][x] = 0

    jump = arr[y][x]
    memo[y][x] += dp(y + jump, x)
    memo[y][x] += dp(y, x + jump)
    return memo[y][x]


if __name__ == '__main__':
    n = int(input())
    arr = [list(map(int, input().split())) for _ in range(n)]
    memo = [[-1] * (n + 1) for _ in range(n + 1)]

    print(dp(0, 0))

File: by-kotlin/kt-boj/solutions/1890-점프.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.math.BigInteger

private lateinit var arr: Array<Array<Int>>
private lateinit var memo: Array<Array<BigInteger>>
private val minusOne = BigInteger.ONE.negate()
private var n = 0

fun main() = with(BufferedReader(InputStreamReader(System.`in`))) {
    n = readLine().toInt()
    arr = Array(n) { Array(n) { 0 } }
    memo = Array(n) { Array(n) { minusOne } }

    for (i in 0 until n)
        arr[i] = readLine().split(" ").map { it.toInt() }.toTypedArray()

    println(dp1890(0, 0))
}

fun dp1890(y: Int, x: Int): BigInteger {
    if (y == n - 1 && x == n - 1) return BigInteger.ONE
    if (y >= n || x >= n) return BigInteger.ZERO

    // arr과 동일한 좌표의 memo 값이 -1이 아니면 이미 지나쳐감
    if (memo[y][x] != minusOne) return memo[y][x]

    memo[y][x] = BigInteger.ZERO
    val jump = arr[y][x]

    memo[y][x] += dp1890(y + jump, x)
    memo[y][x] += dp1890(y, x + jump)

    return memo[y][x]
}
